apps/dashboard/test/models/batch_connect/session_context_test.rb

```diff
@@ -0,0 +1,105 @@
+  end
+  #
+   test "should update cache using update_with_cache" do
+     app = BatchConnect::App.new(router: nil)
+     app.stubs(:form_config).returns(attributes: { num_cores: { widget: "number_field", value: "1" } }, form: ["bc_account", "num_cores"])
+     context = app.build_session_context
+  
+     context.update_with_cache({"bc_account" => "PZS0714", "num_cores" => "28"})
+  
+     assert_equal ["PZS0714", "28"], [context["bc_account"].value, context["num_cores"].value]
+   end
+
+   test "should not update cache if global cache setting disabled" do
+     with_modified_env(OOD_BATCH_CONNECT_CACHE_ATTR_VALUES: 'FALSE') do
+       app = BatchConnect::App.new(router: nil)
+       app.stubs(:form_config).returns(attributes: { num_cores: { widget: "number_field", value: "1" } }, form: ["bc_account", "num_cores"])
+       context = app.build_session_context
+  
+       context.update_with_cache({"bc_account" => "PZS0714", "num_cores" => "28"})
+  
+       assert_equal ["", "1"], [context["bc_account"].value, context["num_cores"].value]
+     end
+   end
+  
+   test "should not update cache if per app cache disabled" do
+     app = BatchConnect::App.new(router: nil)
+     app.stubs(:form_config).returns(cacheable: false, attributes: { num_cores: { widget: "number_field", value: "1" } }, form: ["bc_account", "num_cores"])
+     context = app.build_session_context
+  
+     context.update_with_cache({"bc_account" => "PZS0714", "num_cores" => "28"})
+  
+     assert_equal ["", "1"], [context["bc_account"].value, context["num_cores"].value]
+   end
+  
+   test "should not update single field if per attribute cache disabled" do
+     app = BatchConnect::App.new(router: nil)
+     app.stubs(:form_config).returns(attributes: { bc_account: {value: "PZS0714"}, num_cores: { widget: "number_field", value: "1", cacheable: false } }, form: ["bc_account", "num_cores"])
+     context = app.build_session_context
+     context.update_with_cache({"bc_account" => "PZS0714", :num_cores => 28}  )
+     assert_equal ["PZS0714", "1"], [context["bc_account"].value, context["num_cores"].value]
+   end
+  
+   test "should not update single field if per attribute cache disabled even if global cache setting disabled but per app cache enabled" do
+     with_modified_env(OOD_BATCH_CONNECT_CACHE_ATTR_VALUES: 'FALSE') do
+       app = BatchConnect::App.new(router: nil)
+       app.stubs(:form_config).returns(cacheable: true, attributes: { bc_account: {value: "PZS0714"}, num_cores: { widget: "number_field", value: "1", cacheable: false } }, form: ["bc_account", "num_cores"])
+       context = app.build_session_context
+       cache = { attributes: { num_cores: { widget: "number_field", value: "1", cacheable: false  } } }
+       context.update_with_cache({"bc_account" => "PZS0714", "num_cores" => "28"})
+  
+       assert_equal ["PZS0714", "1"], [context["bc_account"].value, context["num_cores"].value]
+     end
+   end
```

Comment:
I don't think the precedence logic is right. I think it comes from the use of `attribute_cache_disabled?`

This test case shows that. You should change the cached value of `bc_account` here. This test ensures that `num_cores` isn't being cached, but in fact, `bc_account` isn't being cached either we just can't tell because it's the same value in the form, cache and assertion. 

I believe precedence should be defined by specificity. The more specific a setting the higher the precedence. So in this case, `bc_account` _should_ be cacheable because of the app cacheable setting.